lib/ibm_cloud_sdk_core/base_service.rb

```diff
@@ -0,0 +1,275 @@
+  end
+
+  # Initiates the credentials based on the credential file
+  def load_from_credential_file(service_name, separator = "=")
+    credential_file_path = ENV["IBM_CREDENTIALS_FILE"]
+
+    # Home directory
+    if credential_file_path.nil?
+      file_path = ENV["HOME"] + DEFAULT_CREDENTIALS_FILE_NAME
+      credential_file_path = file_path if File.exist?(file_path)
+    end
+
+    # Top-level directory of the project
+    if credential_file_path.nil?
+      file_path = File.join(File.dirname(__FILE__), "/../../" + DEFAULT_CREDENTIALS_FILE_NAME)
+      credential_file_path = file_path if File.exist?(file_path)
+    end
+
+    return if credential_file_path.nil?
+
+    file_contents = File.open(credential_file_path, "r")
+    file_contents.each_line do |line|
+      key_val = line.strip.split(separator)
+      set_credential_based_on_type(service_name, key_val[0].downcase, key_val[1]) if key_val.length == 2
+    end
+  end
+
+  def load_from_vcap_services(service_name:)
+    vcap_services = ENV["VCAP_SERVICES"]
+    unless vcap_services.nil?
+      services = JSON.parse(vcap_services)
+      return services[service_name][0]["credentials"] if services.key?(service_name)
+    end
+    nil
+  end
+
+  def add_default_headers(headers: {})
+    raise TypeError unless headers.instance_of?(Hash)
+
+    headers.each_pair { |k, v| @conn.default_options.headers.add(k, v) }
+  end
+
+  def iam_access_token(iam_access_token:)
+    @token_manager = IAMTokenManager.new(iam_access_token: iam_access_token) if @token_manager.nil?
+    @iam_access_token = iam_access_token
+  end
+
+  def iam_apikey(iam_apikey:)
+    @token_manager = IAMTokenManager.new(iam_apikey: iam_apikey) if @token_manager.nil?
+    @iam_apikey = iam_apikey
+  end
+
+  # @return [DetailedResponse]
+  def request(args)
+    defaults = { method: nil, url: nil, accept_json: false, headers: nil, params: nil, json: {}, data: nil }
+    args = defaults.merge(args)
+    args[:data].delete_if { |_k, v| v.nil? } if args[:data].instance_of?(Hash)
+    args[:json] = args[:data].merge(args[:json]) if args[:data].respond_to?(:merge)
+    args[:json] = args[:data] if args[:json].empty? || (args[:data].instance_of?(String) && !args[:data].empty?)
+    args[:json].delete_if { |_k, v| v.nil? } if args[:json].instance_of?(Hash)
+    args[:headers]["Accept"] = "application/json" if args[:accept_json] && args[:headers]["Accept"].nil?
+    args[:headers]["Content-Type"] = "application/json" unless args[:headers].key?("Content-Type")
+    args[:json] = args[:json].to_json if args[:json].instance_of?(Hash)
+    args[:headers].delete_if { |_k, v| v.nil? } if args[:headers].instance_of?(Hash)
+    args[:params].delete_if { |_k, v| v.nil? } if args[:params].instance_of?(Hash)
+    args[:form].delete_if { |_k, v| v.nil? } if args.key?(:form)
+    args.delete_if { |_, v| v.nil? }
+    args[:headers].delete("Content-Type") if args.key?(:form) || args[:json].nil?
+
+    if @username == "apikey" && !@icp_prefix
```

Comment:
If they use an `icp-` apikey will the SDK authenticate using basicauth `apikey:<icp-apikey>`?

Reply:
it's only for the case when `password` is `icp-xxx`, and `username` is `apikey`.

Reply:
I think we need both. There are services that don't have basicauth (C&C) and we don't generate u/p for those services. In ICP we need to be able to authenticate basicauth if they enter an `icp-` apikey.

Reply:
updated